cuda/context_manager.cpp

```diff
@@ -0,0 +1,258 @@
+#include "context_manager.hpp"
+
+
+bool CUDAContextManager::get_device(hpt_t hpt, CUdevice *pdev) {
+  CUcontext ctx;
+  bool found = get_top_context(hpt, &ctx);
+  if (!found) {
+    return false;
+  }
+  // std::cerr << "get device for thread " << std::get<2>(hpt) << std::endl;
+  auto dev_find = hp_ctx_to_device_.find({std::get<0>(hpt), std::get<1>(hpt), ctx});
+  if (dev_find != hp_ctx_to_device_.end()) {
+    *pdev = dev_find->second;
+    return true;
+  }
+  return false;
+}
+
+bool CUDAContextManager::get_top_context(hpt_t hpt, CUcontext *pctx) {
+  // std::cerr << "get top ctx for thread " << std::get<2>(hpt) << std::endl;
+  auto& hpt_stack = THAPI_AT(hpt_ctx_stack_, hpt);
+  // Note: the primary context is not used automatically, it must be pushed or set
+  // current first and be on the stack
+  if (hpt_stack.empty()) {
+    // std::cerr << "empty stack for thread " << std::get<2>(hpt) << std::endl;
+    return false;
+  }
+  *pctx = hpt_stack.top();
+  return true;
+}
+
+bool CUDAContextManager::get_stream_device(hpt_t hpt, CUstream stream, CUdevice *pdev) {
+  if (stream == 0) {
+    // using the default stream in the top context on the stack
+    return get_device(hpt, pdev);
+  }
+  hp_stream_t hp_stream_key = {std::get<0>(hpt), std::get<1>(hpt), stream};
+  CUcontext ctx;
+  ctx = THAPI_AT(hp_stream_to_ctx_, hp_stream_key);
+  auto ctx_find = hp_ctx_to_device_.find({std::get<0>(hpt), std::get<1>(hpt), ctx});
+  if (ctx_find != hp_ctx_to_device_.end()) {
+    *pdev = ctx_find->second;
+    return true;
+  }
+  return false;
+}
+
+// cuPrimaryCtxRetain_v2_(entry|exit)
+void CUDAContextManager::primary_ctx_retain_entry(hpt_t hpt, CUdevice dev) {
+  // std::cerr << "primaryctx retain entry thread " << std::get<2>(hpt) << std::endl;
+  hpt_entry_dev_[hpt] = dev;
+}
+
+void CUDAContextManager::primary_ctx_retain_exit(hpt_t hpt, CUresult cuResult,
+                                                 CUcontext ctx) {
+  if (cuResultIsError(cuResult)) {
+    return;
+  }
+  // std::cerr << "primaryctx retain exit thread " << std::get<2>(hpt) << std::endl;
+  try {
+    auto entry_dev = THAPI_AT(hpt_entry_dev_, hpt);
```

Comment:
need `&` or not?

Reply:
I think CUdevice, CUcontext, CUfunction, are all pointer types and it doesn't matter, but not sure how I feel about that, is not always obvious. Also we don't mutate them, so even if it had to copy a non small object, it would not break, just be extra copy. Inclined to never use &, definitely need to be consistent here. The one case we need it is for when the value type is a stack and we ARE mutating it... so I guess be consistent of only use it when we really need to mutate.

Reply:
Changed to use auto& for stack and no & everywhere else.

Reply:
Perfect!